File: mgindb/mgindb.php
<?php
require_once 'vendor/autoload.php';
use TextWebSocketClient;

class MginDBClient {
    private $uri;  // WebSocket URI
    private $username;  // Username for authentication
    private $password;  // Password for authentication
    private $websocket;  // WebSocket client

    /**
     * Constructor to initialize the MginDBClient.
     *
     * @param string $host The host address of the MginDB server.
     * @param int $port The port number of the MginDB server.
     * @param string $username The username for authentication.
     * @param string $password The password for authentication.
     */
    public function __construct($host = '127.0.0.1', $port = 6446, $username = '', $password = '') {
        $this->uri = "ws://$host:$port";
        $this->username = $username;
        $this->password = $password;
        $this->websocket = null;
    }

    /**
     * Connects to the MginDB server and authenticates the user.
     *
     * @throws Exception If authentication fails.
     */
    public function connect() {
        $this->websocket = new TextWebSocketClient($this->uri);
        $this->websocket->connect();
        $authData = json_encode(['username' => $this->username, 'password' => $this->password]);
        $this->websocket->send($authData);
        $response = $this->websocket->receive();
        if ($response !== "MginDB server connected... Welcome!") {
            throw new Exception("Failed to authenticate: " . $response);
        }
    }

    /**
     * Sends a command to the MginDB server and returns the response.
     *
     * @param string $command The command to send.
     * @return string The server response.
     */
    public function sendCommand($command) {
        if ($this->websocket === null || !$this->websocket->isConnected()) {
            $this->connect();
        }
        $this->websocket->send($command);
        return $this->websocket->receive();
    }

    /**
     * Sets a value for a given key in the MginDB.
     *
     * @param string $key The key to set.
     * @param string $value The value to set.
     * @return string The server response.
     */
    public function set($key, $value) {
        return $this->sendCommand("SET $key $value");
    }

    /**
     * Manages indices in the MginDB.
     *
     * @param string $action The action to perform (e.g., LIST, ADD).
     * @param string|null $key The key for the action.
     * @param string|null $value The value for the action.
     * @return string The server response.
     */
    public function indices($action, $key = null, $value = null) {
        $command = "INDICES $action" . ($key ? " $key" : '') . ($value ? " $value" : '');
        return $this->sendCommand($command);
    }

    /**
     * Increments a key's value by a given amount in the MginDB.
     *
     * @param string $key The key to increment.
     * @param string $value The amount to increment by.
     * @return string The server response.
     */
    public function incr($key, $value) {
        return $this->sendCommand("INCR $key $value");
    }

    /**
     * Decrements a key's value by a given amount in the MginDB.
     *
     * @param string $key The key to decrement.
     * @param string $value The amount to decrement by.
     * @return string The server response.
     */
    public function decr($key, $value) {
        return $this->sendCommand("DECR $key $value");
    }

    /**
     * Deletes a key from the MginDB.
     *
     * @param string $key The key to delete.
     * @return string The server response.
     */
    public function delete($key) {
        return $this->sendCommand("DEL $key");
    }

    /**
     * Queries the MginDB with a given key and query string.
     *
     * @param string $key The key to query.
     * @param string|null $queryString The query string.
     * @param string|null $options Additional options for the query.
     * @return string The server response.
     */
    public function query($key, $queryString = null, $options = null) {
        $command = "QUERY $key" . ($queryString ? " $queryString" : '') . ($options ? " $options" : '');
        return $this->sendCommand($command);
    }

    /**
     * Counts the number of records for a given key in the MginDB.
     *
     * @param string $key The key to count.
     * @return string The server response.
     */
    public function count($key) {
        return $this->sendCommand("COUNT $key");
    }

    /**
     * Schedules a task in the MginDB.
     *
     * @param string $action The action to perform (e.g., ADD, REMOVE).
     * @param string|null $cronOrKey The cron expression or key for the action.
     * @param string|null $command The command to schedule.
     * @return string The server response.
     */
    public function schedule($action, $cronOrKey = null, $command = null) {
        $cmd = "SCHEDULE $action" . ($cronOrKey ? " $cronOrKey" : '') . ($command ? " $command" : '');
        return $this->sendCommand($cmd);
    }

    /**
     * Subscribes to a key in the MginDB.
     *
     * @param string $key The key to subscribe to.
     * @return string The server response.
     */
    public function sub($key) {
        return $this->sendCommand("SUB $key");
    }

    /**
     * Unsubscribes from a key in the MginDB.
     *
     * @param string $key The key to unsubscribe from.
     * @return string The server response.
     */
    public function unsub($key) {
        return $this->sendCommand("UNSUB $key");
    }

    /**
     * Closes the WebSocket connection to the MginDB server.
     */
    public function close() {
        if ($this->websocket !== null && $this->websocket->isConnected()) {
            $this->websocket->close();
        }
    }
}
?>

File: mgindb/mgindb.py
import json  # Module for JSON operations
import asyncio  # Module for asynchronous programming
import websockets  # WebSocket client library for asyncio
import nest_asyncio  # Module to apply nested asyncio event loop support

# Apply nested asyncio to allow running async code in environments that already have a running event loop
nest_asyncio.apply()

class MginDBClient:
    def __init__(self, protocol='ws', host='127.0.0.1', port=6446, username='', password=''):
        """
        Initializes the MginDBClient with connection parameters.

        Args:
            host (str): The host address of the MginDB server.
            port (int): The port number of the MginDB server.
            username (str): The username for authentication.
            password (str): The password for authentication.
        """
        self.uri = f"{protocol}://{host}:{port}"  # WebSocket URI
        self.username = username  # Username for authentication
        self.password = password  # Password for authentication
        self.websocket = None  # WebSocket connection placeholder

    async def connect(self):
        """
        Asynchronously connects to the MginDB server and authenticates the user.

        Raises:
            Exception: If authentication fails.
        """
        self.websocket = await websockets.connect(self.uri)
        auth_data = {'username': self.username, 'password': self.password}
        await self.websocket.send(json.dumps(auth_data))
        response = await self.websocket.recv()
        if response != "MginDB server connected... Welcome!":
            raise Exception("Failed to authenticate: " + response)

    async def send_command(self, command):
        """
        Asynchronously sends a command to the MginDB server and receives the response.

        Args:
            command (str): The command to send.

        Returns:
            str: The server response.
        """
        if not self.websocket or self.websocket.closed:
            await self.connect()
        await self.websocket.send(command)
        return await self.websocket.recv()

    def _send_sync_command(self, *args):
        """
        Sends a command to the MginDB server synchronously by wrapping it in an asyncio event loop.

        Args:
            *args: The command and its arguments.

        Returns:
            str: The server response.
        """
        loop = asyncio.get_event_loop()
        command_str = ' '.join(str(arg) for arg in args if arg is not None)

        # Special handling for JSON data in SET command
        if 'SET' in command_str and '{' in command_str:
            key, json_data = command_str.split(' ', 2)[1:3]
            try:
                # Validate JSON before sending
                parsed_data = json.loads(json_data)
                json_data = json.dumps(parsed_data)
                command_str = f'SET {key} {json_data}'
            except json.JSONDecodeError:
                print("Invalid JSON data provided.")

        if loop.is_running():
            return asyncio.ensure_future(self.send_command(command_str))
        return loop.run_until_complete(self.send_command(command_str))

    async def close(self):
        """
        Asynchronously closes the WebSocket connection to the MginDB server.
        """
        if self.websocket:
            await self.websocket.close()

    def set(self, key, value):
        """
        Sets a value for a given key in the MginDB.

        Args:
            key (str): The key to set.
            value (str): The value to set.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('SET', key, value)

    def indices(self, action, key=None, value=None):
        """
        Manages indices in the MginDB.

        Args:
            action (str): The action to perform (e.g., LIST, ADD).
            key (str, optional): The key for the action.
            value (str, optional): The value for the action.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('INDICES', action, key, value)

    def incr(self, key, value):
        """
        Increments a key's value by a given amount in the MginDB.

        Args:
            key (str): The key to increment.
            value (str): The amount to increment by.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('INCR', key, value)

    def decr(self, key, value):
        """
        Decrements a key's value by a given amount in the MginDB.

        Args:
            key (str): The key to decrement.
            value (str): The amount to decrement by.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('DECR', key, value)

    def delete(self, key):
        """
        Deletes a key from the MginDB.

        Args:
            key (str): The key to delete.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('DEL', key)

    def query(self, key, query_string=None, options=None):
        """
        Queries the MginDB with a given key and query string.

        Args:
            key (str): The key to query.
            query_string (str, optional): The query string.
            options (str, optional): Additional options for the query.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('QUERY', key, query_string, options)

    def count(self, key):
        """
        Counts the number of records for a given key in the MginDB.

        Args:
            key (str): The key to count.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('COUNT', key)

    def schedule(self, action, cron_or_key=None, command=None):
        """
        Schedules a task in the MginDB.

        Args:
            action (str): The action to perform (e.g., ADD, REMOVE).
            cron_or_key (str, optional): The cron expression or key for the action.
            command (str, optional): The command to schedule.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('SCHEDULE', action, cron_or_key, command)

    def sub(self, key):
        """
        Subscribes to a key in the MginDB.

        Args:
            key (str): The key to subscribe to.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('SUB', key)

    def unsub(self, key):
        """
        Unsubscribes from a key in the MginDB.

        Args:
            key (str): The key to unsubscribe from.

        Returns:
            str: The server response.
        """
        return self._send_sync_command('UNSUB', key)